hash/hash_comparator.go

```diff
@@ -132,12 +141,42 @@ func CompareScrypt(_ context.Context, password []byte, hash []byte) error {
 	return errors.WithStack(ErrMismatchedHashAndPassword)
 }
 
+func CompareSSHA(_ context.Context, password []byte, hash []byte) error {
+
+	if _, err := ssha.Validate(string(password), string(hash)); err != nil {
+		return errors.WithStack(ErrMismatchedHashAndPassword)
+	}
+
+	return nil
+}
+
+func CompareSSHA256(_ context.Context, password []byte, hash []byte) error {
+
+	if _, err := ssha256.Validate(string(password), string(hash)); err != nil {
+		return errors.WithStack(ErrMismatchedHashAndPassword)
+	}
+
+	return nil
+}
+
+func CompareSSHA512(_ context.Context, password []byte, hash []byte) error {
+
+	if _, err := ssha512.Validate(string(password), string(hash)); err != nil {
+		return errors.WithStack(ErrMismatchedHashAndPassword)
+	}
+
+	return nil
+}
+
 var (
 	isBcryptHash   = regexp.MustCompile(`^\$2[abzy]?\$`)
 	isArgon2idHash = regexp.MustCompile(`^\$argon2id\$`)
 	isArgon2iHash  = regexp.MustCompile(`^\$argon2i\$`)
 	isPbkdf2Hash   = regexp.MustCompile(`^\$pbkdf2-sha[0-9]{1,3}\$`)
 	isScryptHash   = regexp.MustCompile(`^\$scrypt\$`)
+	isSSHAHash     = regexp.MustCompile(`^{SSHA}.*`)
```

Comment:
Why are we not follwing the $-separator format here?

Reply:
SHA and SSHA are often stored with, for example, {SSHA512} prepending the value.  An example function being something like this:

`$hash = "{SSHA256}" . base64_encode(hash('sha256', $password . $salt) . $salt);`

This is pretty standard for these methods.  We could change this but that would burden users with transforming their inputs when importing users.

I'm happy to make the change to use $SSHA$ where appropriate if everyone prefers that approach.

Reply:
I guess the question is if kratos wants to support only SSHA or also custom implementations (for example reversed salt and password order).

Reply:
Could you please provide a source for this statement?

I still think that we'll should use the PHC format here, because SHA for passwords is not standardized and many implementations with different salting mechanisms exist. So I must insist that this importer follows suit here. We can always write a CLI helper (`grep` :D) which converts whatever SHA format into PHC.

You can take a look at the MD5 PR ( #2725 ) to see how we solved it there!

Reply:
@aeneasr I think the confusion here is, that SSHA (and SSHA256, and SSHA512) is what @zambien wanted to implement with this PR in the first place and that is standardized (e.g. https://community.canvaslms.com/t5/SIS-User-Articles/SSHA-Password-Generation/ta-p/243730) and you and @serlo for example would prefer a more flexible implementation of all sorts of SHA approaches.

Reply:
The other thread (https://github.com/ory/kratos/pull/2741#discussion_r980357651) is more up to date on this.

Reply:
> @aeneasr I think the confusion here is, that SSHA (and SSHA256, and SSHA512) is what @zambien wanted to implement with this PR in the first place and that is standardized (e.g. https://community.canvaslms.com/t5/SIS-User-Articles/SSHA-Password-Generation/ta-p/243730) and you and others (e.g we at @serlo) would prefer a more flexible implementation for all sorts of SHA approaches.

I've started on some code which will implement both SHA and SSHA and also will dynamically handle the presence or absence of a salt based on the comments in the various threads.  Been a busy couple weeks but I'll commit it once I feel it is close to ready to be reviewed.

Reply:
Ok, for SSHA I think it makes sense to use the format as standardized `{SSHA}...`! For SSH as discussed with the PHC string format :) Sorry for confusing this here a bit.